tools/whodraws.py

```
@@ -0,0 +1,127 @@
+    smin = np.take_along_axis(S, ordS[:, :1], 1)[:, 0]
+    smid = np.take_along_axis(S, ordS[:, 1:2], 1)[:, 0]
+    axmin = np.take_along_axis(R, ordS[:, None, :1], 2)[:, :, 0]
+    axmax = np.take_along_axis(R, ordS[:, None, 2:], 2)[:, :, 0]
+    vmin, vmax_ = np.abs(axmin[:, 1]), np.abs(axmax[:, 1])
+    needle = smid / np.maximum(smax, 1e-9) < 0.5
+    blob = (~needle) & (smin / np.maximum(smax, 1e-9) > 0.5)
+    plate = ~needle & ~blob
+
+    # 局所地面
+    vote = live & (smax < 0.15) & (op > 0.5)
+    lo = P[vote][:, [0, 2]].min(0); hi = P[vote][:, [0, 2]].max(0)
```

Ruff (0.16.3)

[error] Multiple statements on one line (semicolon)

(E702)